ERAP updater using palletjack

jacobdadams opened the pull request:
Updates the ERAP feature service and map using github.com/agrc/palletjack for the heavy lifting.

jacobdadams committed:
chore: first setup

jacobdadams committed:
feat: update functions from prototyping

jacobdadams committed:
chore: supervisor and initial framework

jacobdadams committed:
wip: data coupling classes

jacobdadams committed:
fix: use an arcpy-friendly pytest

jacobdadams committed:
feat: update from data_dict

jacobdadams committed:
feat: mocking arcpy cursors!!!

jacobdadams committed:
feat: more tests and fixes

jacobdadams committed:
feat: ColorRampReclassifier

jacobdadams committed:
wip: rough, untested main script

jacobdadams committed:
wip: rearrange process, move more to secrets

jacobdadams committed:
feat: logging

jacobdadams committed:
chore: ignore knownhosts

jacobdadams committed:
chore: various fixes to make it functional

jacobdadams committed:
fix: return a blank dataframe for len, shape calls

jacobdadams committed:
feat: get summaries from steps, send email

jacobdadams committed:
feat: new test for column check

jacobdadams committed:
feat: move logger, supervisor to init

jacobdadams committed:
feat: save to log file

jacobdadams committed:
fix: on second thought ... tis a silly place

jacobdadams committed:
feat: use palletjack

stdavis commented on .vscode/settings.json:
```diff
@@ -34,5 +34,6 @@
   "python.testing.pytestEnabled": true,
-  "python.testing.unittestEnabled": false
+  "python.testing.unittestEnabled": false,
+  "python.pythonPath": "C:\\Users\\jdadams\\Miniconda3\\envs\\arcpy\\python.exe"
```
I thought that VSCode did away with this setting at the project level?

steveoh commented on setup.py:
```diff
@@ -33,7 +33,8 @@
     install_requires=[
-        # 'package==1.0.*'
+        'agrc-supervisor==2.0.*',
+        'pysftp==0.2.9',
```
**question**: should pysftp be a dependency of palletjack? Is pallet jack a local module then?

jacobdadams replied:
Good catch, leftover from before splitting out palletjack.

jacobdadams committed:
chore: PR suggestions

jacobdadams committed:
fix: u instead of a

steveoh commented on setup.py:
```diff
@@ -10,11 +10,11 @@
+    description='Updates for the Emergency Rental Assistance map',
+    author='UGRC',
     author_email='agrc@utah.gov',
     url='https://github.com/agrc/python',
```
**nitpick**: this url needs to change

steveoh commented:
I think it would be nice to update the readme and make the github action successful in this pull request.

jacobdadams committed:
chore: try removing testing from action

jacobdadams committed:
chore: readme, setup.py updates

jacobdadams committed:
chore: extend and document secrets

jacobdadams committed:
chore: readme updates

jacobdadams commented:
Moved to draft pending palletjack completion.

Also, re: CI testing, I'm running into the same problem as with Auditor a while ago. ERA imports palletjack, which imports arcpy, which can't be installed in the CI environment. I've modified the action to just build, no tests. Are you aware of any better work arounds?

steveoh commented:
mocking would probably be the best solution. try except the import and then supply a mock in the exception handler.

jacobdadams committed:
chore: fix ci testing

jacobdadams committed:
feat: rotate data folders

jacobdadams committed:
chore: streamline folder rotation

jacobdadams committed:
chore: docs

jacobdadams committed:
chore: missed docs

jacobdadams committed:
fix: folder deletion; slicing is hard

jacobdadams committed:
chore: docs

jacobdadams committed:
chore: clarity

jacobdadams committed:
chore: use FileRotator, update logging

jacobdadams committed:
chore: update palletjack dep

jacobdadams commented:
Ok, now that palletjack is up we can push this out. I've implemented a folder rotator to rotate the data folder as needed; I would appreciate a review of this (rotator.py). 

Test coverage is low because the tests are focused around the individual methods themselves. The main script just integrates supervisor, palletjack, and rotator.py, which are all individually tested.

steveoh committed:
chore: ignore secrets template

steveoh commented on src/era/main.py:
```diff
@@ -0,0 +1,122 @@
+from supervisor.message_handlers import SendGridHandler
+from supervisor.models import MessageDetails, Supervisor
+
+from . import rotating, secrets
```
**thought**: there are a few of properties off of the secrets that don't appear to be secrets. Having them in source control would make development simpler for others.

jacobdadams replied:
I tried to leave everything I thought I could safely leave in `secrets_template.py`. Were you thinking of the service URLs/fieldnames/itemids? My thought was to leave them out because they could change, but I could copy them over to secrets_template.py if you think it's ok to have those in version control.

I guess my main thinking is that secrets.py kind of became a general settings file instead of breaking it out into a separate secrets-only file and a more general settings file. I prefer the simplicity of a single file, but I'm open to reasons why that's not the best idea.

steveoh replied:
My suggestion is to separate secrets from configuration.

steveoh commented on src/era/rotating.py:
```diff
@@ -0,0 +1,106 @@
+        try:
+            download_dir_path.mkdir(exist_ok=exist_ok)
+        except FileNotFoundError as error:
+            raise FileNotFoundError(f'Base directory `{self.base_dir}`` does not exist.') from error
```
**suggestion**: is this an extra back tick?

**question**: The raise from is new to me. Does that merge the two errors when re raising?

```suggestion
            raise FileNotFoundError(f'Base directory `{self.base_dir}` does not exist.') from error
```

jacobdadams replied:
Thanks for catching the back tick. And that is python's exception chaining (https://docs.python.org/3/reference/simple_stmts.html#the-raise-statement, https://www.python.org/dev/peps/pep-3134/). It raises the `from` exception and then says it was the direct cause of the manually raised exception. I'm using it here to display more info about what I'm pretty sure would be the only cause of the original error.

steveoh commented on src/era/rotating.py:
```diff
@@ -0,0 +1,106 @@
+                'no folders deleted'
+            )
+            max_folder_count = 0
+        return sorted(folder_paths)[:-max_folder_count]
```
**question**: `folder_paths` is a `Path` object in this place right. Does it sort based on string semantics?  Would you ever want to use os level folder metadata for knowing the age of a folder or could that be problematic?

jacobdadams replied:
AFIU, it sorts based on lexicographic order, which means it takes into account the parent directories before sorting purely by alphabetical order. 

I've always shied away from trusting filesystem-level metadata for dates because I've no guarantee that some other process or user hasn't inadvertently changed the attribute I'm depending on. It's probably not that likely or hazardous here, but it's just kind of my default to manually encode that info in the filename. That way I/someone else would have to be very deliberate in changing it.

The broader implication of the `date_format` and regex `pattern` is intended to ensure any folder deletion has as narrow a scope as possible. If someone put another folder in the directory, it's an unlikely pattern for human-named folders so it won't grab it "by accident". Also, it allows someone to make a copy or rename a folder they want to keep around for testing or data archiving by changing the folder name.  

That's my thought process, anyways.

steveoh commented on src/era/rotating.py:
```diff
@@ -0,0 +1,106 @@
+                self._class_logger.debug(f'Attempting to delete `{folder}`')
+                shutil.rmtree(folder)
+            except Exception:
+                pass
```
**question**: should the folders that can't be deleted get logged for manual intervention? Couldn't they foul up the next run when calling `_get_all_but_n_most_recent_folder_paths`

jacobdadams replied:
I see what you mean... I put in the attempt log so that if you noticed it wasn't deleting folders as expected you could track it down with DEBUG level, but if each run introduced a new "sticky" folder eventually you'd get to the point where you're not deleting any folders and the directory could grow out of control. The script should create a new folder each time, so the only impact would be having too many files (I've seen very chatty logs [looking at you, samba] fill all available disk space and crash a server, but we're talking hundreds of lines of logging per second). 

Maybe it'd be best to log it as a warning or an error here instead of silently passing?

steveoh replied:
I think a warning level log would be great here.

steveoh commented on src/era/rotating.py:
```diff
@@ -0,0 +1,106 @@
+                Defaults to ''.
+            date_format (str, optional): Date format argument to strftime(). Defaults to '%Y%m%d_%H%M%S'
+                (YYYYMMDD_hhmmss).
+            exist_ok (bool, optional): Overwrite an existing folder?. Defaults to False.
```
I just copied the docs from pathlib here

```suggestion
            exist_ok (bool, optional): If exist_ok is false (the default), FileExistsError is raised if the target directory already exists. Defaults to False.
```

jacobdadams replied:
Good call, that's more useful. I've added it and made it a little more specific to the script.

steveoh commented on src/era/rotating.py:
```diff
@@ -0,0 +1,106 @@
+        self._class_logger.debug(f'Attempting to create new directory `{download_dir_path}`')
+        try:
+            download_dir_path.mkdir(exist_ok=exist_ok)
+        except FileNotFoundError as error:
```
**question**: did you mean for this to be a `FileExistsError` or to add the `parents` [parameter](https://docs.python.org/3/library/pathlib.html#pathlib.Path.mkdir)?

jacobdadams replied:
I meant to leave `parents` at the default `False` and raise an error that gave a hint more than just a `FileNotFound` error. Personally, I'd rather it fail and warn me than write output to a place I wasn't expecting.

steveoh replied:
Ok, so you are ok with the script exiting here or with a FileExistsError since supervisor will catch it and notify us

jacobdadams replied:
Exactly, this directory needs to exist for the download to work, so it should bomb out with the appropriate notification.

steveoh commented on src/era/rotating.py:
```diff
@@ -0,0 +1,106 @@
+        INFO:
+            Directories deleted as part of rotation
+
+        Raises FileNotFoundError if base_dir does not exist. Silently swallows any errors when deleting folders
```
**thought**: this would also raise FileExistsError if `exists_ok` was false

steveoh commented on src/era/rotating.py:
```diff
@@ -0,0 +1,106 @@
+        folder_paths_to_delete = self._get_all_but_n_most_recent_folder_paths(prefix, pattern, max_folder_count)
+        deleted_folders = self._delete_old_folders(folder_paths_to_delete)
+        self._class_logger.info(
+            f'Deleted {len(deleted_folders)} folder(s) for rotation: {[str(folder) for folder in deleted_folders]}'
```
**suggestion**: using the `folder.name` might make the logs a little less dense since the `base_dir` is static

jacobdadams replied:
It would, but I'm inclined to stay with the full path because it explicitly tells you which folders were deleted, with no room for misinterpretation or hidden bugs where somehow `base_dir` changes to something the user wasn't expecting.

steveoh commented on src/era/secrets_template.py:
```diff
@@ -0,0 +1,52 @@
+
+#: AGOL Account Settings
+AGOL_ORG = 'https://utah.maps.arcgis.com'
+AGOL_USER = ''
+AGOL_PASSWORD = ''
```
**question**: are we able to register this script as an app in agol and use client id and secret for authentication tokens rather than a real account?

jacobdadams replied:
I've not done that before, I'll have to look into it. It makes sense.

jacobdadams committed:
chore: code review suggestions/updates

jacobdadams committed:
suggestion: warn if folder can't be deleted

jacobdadams committed:
feat: arcgis-only palletjack

jacobdadams committed:
chore: more dependency breathing room

jacobdadams commented:
Converting to draft while working on google cloud function adventure.

steveoh commented:
> Converting to draft while working on google cloud function adventure.

I might be best to finalize this this pr and open a new one for the function stuff.

jacobdadams commented:
I think you're right... it's starting to get messy, branches on branches. Two steps forward and one step back! Riders, away!

jacobdadams committed:
chore: remove vestigial arcpy
